xds/src/main/java/io/grpc/xds/RingHashLoadBalancer.java

```diff
@@ -93,81 +95,198 @@ public boolean acceptResolvedAddresses(ResolvedAddresses resolvedAddresses) {
+   * </ol>
+   */
+  @Override
+  protected void updateOverallBalancingState() {
+    checkState(!getChildLbStates().isEmpty(), "no subchannel has been created");
+    boolean startConnectionAttempt = false;
+
+    // Calculate the current overall state and whether we should attempt a connection
+    int numIdle = 0;
+    int numReady = 0;
+    int numConnecting = 0;
+    int numTransientFailure = 0;
+    for (ChildLbState childLbState : getChildLbStates()) {
+      ConnectivityState state = childLbState.getCurrentState();
+      if (state == READY) {
+        numReady++;
+        break;
+      } else if (state == TRANSIENT_FAILURE) {
+        numTransientFailure++;
+      } else if (state == CONNECTING ) {
+        numConnecting++;
+      } else if (state == IDLE) {
+        numIdle++;
+      }
     }
-    // If we need to proactively start connecting, iterate through all the subchannels, starting
-    // at a random position.
-    // Alternatively, we should better start at the same position.
-    connectionAttemptIterator = subchannels.values().iterator();
-    int randomAdvance = random.nextInt(subchannels.size());
-    while (randomAdvance-- > 0) {
-      connectionAttemptIterator.next();
+    ConnectivityState overallState;
+    if (numReady > 0) {
+      overallState = READY;
+    } else if (numTransientFailure >= 2) {
+      overallState = TRANSIENT_FAILURE;
+      startConnectionAttempt = (numConnecting == 0);
+    } else if (numConnecting > 0) {
+      overallState = CONNECTING;
+    } else if (numTransientFailure == 1 && getChildLbStates().size() > 1) {
+      overallState = CONNECTING;
+      startConnectionAttempt = true;
+    } else if (numIdle > 0) {
+      overallState = IDLE;
+    } else {
+      overallState = TRANSIENT_FAILURE;
     }
 
-    // Update the picker before shutting down the subchannels, to reduce the chance of race
-    // between picking a subchannel and shutting it down.
-    updateBalancingState();
-    for (Subchannel subchann : removedSubchannels) {
-      shutdownSubchannel(subchann);
+    RingHashPicker picker = new RingHashPicker(syncContext, ring, getImmutableChildMap());
+    // TODO(chengyuanzhang): avoid unnecessary reprocess caused by duplicated server addr updates
+    getHelper().updateBalancingState(overallState, picker);
+    this.currentConnectivityState = overallState;
+
+    // While the ring_hash policy is reporting TRANSIENT_FAILURE, it will
+    // not be getting any pick requests from the priority policy.
+    // However, because the ring_hash policy does not attempt to
+    // reconnect to subchannels unless it is getting pick requests,
+    // it will need special handling to ensure that it will eventually
+    // recover from TRANSIENT_FAILURE state once the problem is resolved.
+    // Specifically, it will make sure that it is attempting to connect to
+    // at least one subchannel at any given time.  After a given subchannel
```

Comment:
This comment paragraph corresponds to this paragraph in A61
```
As per [gRFC A42](https://github.com/grpc/proposal/blob/28ed4fbd237266a10c99654583065177c777292c/A42-xds-ring-hash-lb-policy.md), the ring_hash policy normally requires...
```

The `connectionAttemptIterator ` was used to achieve that. Now this needs to be removed and change to the psuedo-code mentioned in grpc.

Reply:
Removed connectionAttemptIterator.